Intrinsic support for multiple backends and general restyling

msessa opened the pull request:
Hi @djenriquez @alexunwin 

Note: I'm opening a PR although there is still a bit of work to do here before it can be merged, but this way is easier to discuss changes.

This PR shifts the discovery of secret backends and authentication backend (post-login) to the Menu component, which dynamically renders a nested list of mounted backends.
I believe vault-ui needs to be designed following the core architecture of vault, which allow multiple instances of a backend to be mounted on arbitrary paths in the logical filesystem.
Also I believe it's not the responsibility of the individual backend-interface components to discover their namespaces.

This should make it easier to develop support for other types of secret backends (aws, mysql etc) and authentication backends (userpass, ldap, aws, etc)

Edit: To demonstrate the point above I've moved the GitHub and EC2 components from Policies to the appropriate location for the new model.

You can test all this with
```
./run-docker-compose-dev
```

A breakdown of the changes:
- Redesigned left side menu navigation using material-ui components
- Menu entries are dynamically loaded if token has read access to `/sys/mounts` and `/sys/auth` otherwise it will fallback to the standard view
- Refactored Secrets page to accommodate this approach
- Redesigned Secrets browser with material-ui components
- Refactored Policies with new layout

It looks like this:
<img width="1440" alt="screen shot 2017-02-02 at 10 07 28 pm" src="https://cloud.githubusercontent.com/assets/1912143/22547084/0b1f25e0-e994-11e6-92d6-68f4032ea2c9.png">

Commit:
Begin role management feature

Commit:
First commit dynamic navigation

Commit:
Refactor secret management

Commit:
Support dynamic auth backends

Commit:
Layout changes

Commit:
Various component restructuring and refactoring

Commit:
mount github in docker-compose env

Commit:
fix bug in listing after new object creation

Commit:
mount aws-ec2 in docker-compose env

Commit:
remove bad hack

Commit:
Role listing and editing

Commit:
Fix bugs

Commit:
Fixes in snackbar messages

djenriquez commented:
Looks fantastic so far @msessa-cotd. I'll dive deeper into the details later tonight, awesome work!

Commit:
Complete implementation of various role functions

Commit:
Fix unexpected behaviour when using referenced objects

Commit:
First commit policy picker component

Commit:
Merge branch 'master' into token_management

msessa commented:
Hi @djenriquez, did you have a chance to take a look? Any thoughts?

djenriquez commented:
@msessa-cotd sorry for the delay. I haven't had a chance to take a look yet. I'll make time tonight to do so.

djenriquez commented:
@msessa-cotd I'm seeing the added keys you're using to figure out the permissions for secrets navigation, but I do not see the same menu items as the screen shot above. Did you decide to move back on this?

msessa commented:
Did you run the `run-docker-compose-dev` in my branch? I've added a few lines there to mount a second generic backend and create some objects

djenriquez commented:
Yup, ran `docker-compose build` also, which is probably a step we should add to the doc if devs are switching branches with different dependencies:
```
vault-ui $ git branch
* dynamic_nav

vault-ui $ . run-docker-compose-dev 
------------- docker-compose up -d -------------
Creating network "vaultui_default" with the default driver
Creating vaultui_vault_1
Creating vaultui_vault-ui_1

------------- docker-compose ps -------------
      Name             Command             State              Ports       
-------------------------------------------------------------------------
vaultui_vault-     npm run serve      Up                 0.0.0.0:8000->80 
ui_1                                                     00/tcp           
vaultui_vault_1    docker-            Up                 0.0.0.0:8200->82 
                   entrypoint.sh                         00/tcp           
                   serve ...                                              

------------- vault auth a162c563-6f2f-31ff-6294-2a524194bee3 -------------
Successfully authenticated! You are now logged in.
token: a162c563-6f2f-31ff-6294-2a524194bee3
token_duration: 0
token_policies: [root]

------------- vault status -------------
Sealed: false
Key Shares: 1
Key Threshold: 1
Unseal Progress: 0
Version: 0.6.4
Cluster Name: vault-cluster-773d6d89
Cluster ID: e2779473-4af4-f927-5469-3250646ecd91

High-Availability Enabled: false

------------- vault auth-enable userpass -------------
Successfully enabled 'userpass' at 'userpass'!

------------- vault auth-enable github -------------
Successfully enabled 'github' at 'github'!

------------- vault auth-enable -path=awsaccount1 aws-ec2 -------------
Successfully enabled 'aws-ec2' at 'awsaccount1'!

------------- vault policy-write admin /app/admin.hcl -------------
Policy 'admin' written.

------------- vault write auth/userpass/users/test password=test policies=admin -------------
Success! Data written to: auth/userpass/users/test

------------- vault write secret/test somekey=somedata -------------
Success! Data written to: secret/test

------------- vault mount -path=ultrasecret generic -------------
Successfully mounted 'generic' at 'ultrasecret'!

------------- vault write ultrasecret/moretest somekey=somedata -------------
Success! Data written to: ultrasecret/moretest

------------- vault write ultrasecret/dir1/secret somekey=somedata -------------
Success! Data written to: ultrasecret/dir1/secret

------------- vault write ultrasecret/dir2/secret somekey=somedata -------------
Success! Data written to: ultrasecret/dir2/secret

------------- vault write ultrasecret/dir2/secret2 somekey=somedata -------------
Success! Data written to: ultrasecret/dir2/secret2

------------- vault write ultrasecret/admincantlistthis/butcanreadthis somekey=somedata -------------
Success! Data written to: ultrasecret/admincantlistthis/butcanreadthis

------------- vault write ultrasecret/admincantreadthis somekey=somedata -------------
Success! Data written to: ultrasecret/admincantreadthis

------------- Vault Root Token -------------
vault_1     | Root Token: a162c563-6f2f-31ff-6294-2a524194bee3
```

<img width="950" alt="screen shot 2017-02-08 at 7 14 56 pm" src="https://cloud.githubusercontent.com/assets/4636740/22767850/f8714cfe-ee32-11e6-81cf-825367ecf2dc.png">

Commit:
Re enable webpack

msessa commented:
Ooops I commented out webpack in docker-compose.yml and forgot to revert it before I committed.

Pull now

djenriquez commented:
The left hand menu is definitely looking better. The core real estate seems to have regressed though with some views looking like this: 

<img width="1007" alt="screen shot 2017-02-08 at 7 59 55 pm" src="https://cloud.githubusercontent.com/assets/4636740/22768790/333acba2-ee39-11e6-85cd-d3befa421400.png">

I think if we maximize the real estate to somehow remove the "modal", it'll present itself better.

The new distinction between namespaces and actual secrets is great: 
<img width="859" alt="screen shot 2017-02-08 at 8 02 08 pm" src="https://cloud.githubusercontent.com/assets/4636740/22768843/917c3872-ee39-11e6-9f73-bb38ac45d3ad.png">

However, the new navigation bar for secrets seems unintuitive since users still enter namespaced keys with `/` but the display shows arrows: 
<img width="315" alt="screen shot 2017-02-08 at 8 00 27 pm" src="https://cloud.githubusercontent.com/assets/4636740/22768821/6182ffc0-ee39-11e6-8cb2-709dc83f76ef.png">

vs

<img width="332" alt="screen shot 2017-02-08 at 8 01 37 pm" src="https://cloud.githubusercontent.com/assets/4636740/22768826/6c354888-ee39-11e6-9301-ad5357ae84f1.png">

Lastly, the use of the toastbar is exactly what I visioned to help provide better feedback to the user. I also like the experience of denied access to secrets and name spaces. Nice stuff!

Do you agree with my feedback on the navigation bar and core real estate?

Commit:
Merge token management

Commit:
preview policy picker

msessa commented:
I agree. I'll make the breadcrumb nav better in the secret browser. 
Also some of the views like secret wrapping, github, EC2 need to be retouched to match the new styling (the one I've done for secrets and policies).

I've also designed a "policy picker" component since choosing from existing policies is a very common task in vault. You can check it out if you pull now and go in Token/Create New Token or Role. It looks like this:
![image](https://cloud.githubusercontent.com/assets/1912143/22769325/1f90ed0e-eedc-11e6-914e-ace4062304b0.png)


Also I want to extend all the components to be addressable by URL like the secrets page.
For example you should be able to point to
`https://vaultui/system/policies/policyname`
and the policy editor should pop up.

Anyway if you like what's been done so far, I can continue working on it and finish the remaining bits, once we merge this branch I will start writing new components for other secret/auth backends.

I'll have to prioritise the ones my company uses most (EC2 auth, RADIUS auth, IAM creds) but eventually I'd like to see interfaces for all supported backend.

djenriquez commented:
Thats fantastic, and yes, prioritize what you need. When things settle a bit with my new job I'll have time to be more active, but really appreciate the work!

Commit:
Check current token validity using lookup-self

Commit:
Revert "Check current token validity using lookup-self"

This reverts commit c9554de2d11f5791dcb56906c61284e52455d213.

Commit:
Revert "preview policy picker"

This reverts commit f2075bb53fc4836d40fdf0b378d13b9438567306.

Commit:
First commit enhanced app header

Commit:
URL addressable policies

Commit:
Restyle breadcrumb nav

Commit:
Better error handling

Commit:
earlier versions of vault return 400 instead of 403

Commit:
Add babel eslint parser

Commit:
Redesign welcome screen

When logging in with a token that doesnt have permissions
to list auth/secret backends, a warning message is displayed
on the welcome screen, informing the user of the missing permissions
and how to fix it

Commit:
set returnto querystring for post-login redirection

Commit:
Integrate wrapping in generic secret backend

Commit:
eslint cleanups

Commit:
ResponseWrapper integration with shared Wrapper component

Commit:
Consolidate layout for settings and github components

Commit:
Integrate policypicker in token backend component

Commit:
add test unprivileged user

msessa commented:
Hi @djenriquez,

I've completed most of the restyling around the app. Would you be able to take a look?

There's also a few new substantial changes:
- Validity of token is now checked on app mount
- App header refactored and extended with current identity information
![image](https://cloud.githubusercontent.com/assets/1912143/23094738/98083d00-f652-11e6-8730-0325ec49f7d4.png)
- Policies are now URL addressable
- Welcome screen now displays warning messages if backends cant be listed (`./run-docker-compose-dev` and login with `u:lame p:lame` to preview)
- Preserve the original path when unauthenticated users are redirected to login
- New generic `SecretWrapper` component. This component can be reused in the app to wrap any type of data.
- Integrate `SecretWrapper` in generic secret component
- Integrate `SecretWrapper` in response wrapping page (renamed to Data Wrapper)
- `PolicyPicker` integration in token/role creation dialog

The `SecretWrapper` component is particularly interesting because it's coupled with a very basic router endpoint `/unwrap` which displays a readonly json editor with the unwrapped data. It doesn't require any authentication to vault UI.

Example wrapping an existing secret:
![image](https://cloud.githubusercontent.com/assets/1912143/23094865/b3463e8a-f654-11e6-90ae-8f4f5dcf380b.png)
![image](https://cloud.githubusercontent.com/assets/1912143/23094869/c199416c-f654-11e6-9931-7fb5bbf0ce02.png)

The provided URL can be used with no prior authentication to display the wrapped secret.

Let me know your thoughts. 

Cheers

djenriquez commented:
@msessa-cotd 

Without looking at the code, the experience is most definitely improved, _absolutely_ awesome stuff! The token, username and connected Vault address is great information for the user to have. Response wrapping also has a better experience now.

The only thing I find questionable is the modal style view of the core real estate. I believe it would look and feel better if it was maximized somehow. Ideally, the display would take up the entire core space without floating over a white background. Was there a reason for the floating-style displays vs just using the space?

<img width="846" alt="screen shot 2017-02-18 at 3 23 22 pm" src="https://cloud.githubusercontent.com/assets/4636740/23097673/3622e6b2-f5ee-11e6-83f2-d8e69c6334b9.png">

Also for language/spelling, let's stick with US English to remain consistent with the Vault project itself. (Customize vs customize, Preferences vs Preferencies, etc)

None-the-less, bravo 🎉!

Commit:
small layout fixes

msessa commented:
hi @djenriquez 

I've reclaimed some space on the page and taken away most of the depth effect.

Also fixed those australian spellings ;) sorry, force of habit

djenriquez commented:
Looks much better. I think you missed one with the Tokens page. 
![screen shot 2017-02-20 at 9 38 08 am](https://cloud.githubusercontent.com/assets/4636740/23136175/50fef2f6-f750-11e6-8ae1-1f3e9c25e7ad.png)

Otherwise, LGTM 👍

djenriquez commented:
After these two PRs are merged in, PTAL at [Release Draft 1.1.0](https://github.com/djenriquez/vault-ui/releases/tag/untagged-793e8a5b74c39298f443) and make sure the release notes are accurate. I think you should have access to that.

Commit:
Remove unused nodejs code

Commit:
Fix tokens component layout

msessa commented:
Fixed layout for the tokens page.

Can you guys consider merging?
